examples/solvers/chambolle_pock_denoising_tgv.py

```diff
@@ -0,0 +1,134 @@
+"""Total Generalized Variation denoising using the Chambolle-Pock solver.
+
+Solves the optimization problem
+
+    min_x ||x - d||_2^2 + alpha TGV_2(x)
+
+Where ``d`` is given noisy data TGV_2 is the second order total generalized
+variation of ``x``, defined as
+
+    TGV_2(x) = min_y ||Gx - y||_1 + beta ||Ey||_1
+
+where ``G`` is the spatial gradient operating on the scalar-field ``x`` and
+``E`` is the symmetrized gradient operating on the vector-field ``y``.
+Both one-norms ||.||_1 take the one-norm globally and the two-norm locally,
+i.e. ||y||_1 := sum_i sqrt(sum_j y_i(j)^2) where y_i(j) is the j-th value
+of the vector y_i at location i.
+
+The problem is rewritten as
+
+    min_{x, y} ||x - d||_2^2 + alpha ||Gx - y||_1 + alpha * beta ||Ey||_1
+
+which can then be solved with the Chambolle-Pock method.
+
+References
+----------
+[1] K. Bredies and M. Holler. A TGV-based framework for variational image
+decompression, zooming and reconstruction. Part II: Numerics. SIAM Journal
+on Imaging Sciences, 8(4):2851-2886, 2015.
+"""
+
+import numpy as np
+import odl
+
+# --- Set up the forward operator (identity) --- #
+
+# Reconstruction space: discretized functions on the rectangle
+# [-20, 20]^2 with 300 samples per dimension.
+n = 300
+U = odl.uniform_discr(min_pt=[-20, -20], max_pt=[20, 20], shape=[n, n],
+                      dtype='float32')
+
+# Create the forward operator
+A = odl.IdentityOperator(U)
+
+# --- Generate artificial data --- #
+
+# Create phantom
+def phantom_func(pts):
+    x, y = pts
+    sign = ((x < -10) | (x > 10) | (y < -10) | (y > 10)) * 3 - 2
+    return (x / 20.0) * sign
+
+phantom = U.element(phantom_func)
+phantom.show(title='Phantom')
+
+# Create sinogram of forward projected phantom with noise
+data = A(phantom)
+data += odl.phantom.white_noise(A.range) * np.mean(data) * 0.1
```

Comment:
I changed the phantom to be between 0 and 1 again (which is what I had in mind originally). Then, the abs is not necessary.